### PR TITLE
Plugin system

### PR DESCRIPTION
**Pull request by mariots**
Solves #8 This PR includes an updated plugin engine architecture and allowing the engine to register callback functions. The callback functions will request certain streams and may be called with the streams as arguments.

### CODE REVIEW

File: plugins/fakePlugin.py [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
 
 The class name must be the same as the file name.
 """
-class FakePlugin():
+class FakePlugin(PluginBase.PluginBase):
 
-    def __init__(self):
-        return
+    def init(self):
+        self.pluginEngine.RegisterCallback("realtime", "HelloWorld", self.HelloWorld, ["cameraStills", "otherStream"])
```
Comment (alexweav) [inline plugins/fakePlugin.py]: we probably don't need this to pass the strings "realtime" and "HelloWorld." we're probably not gonna get to the historical engine (and that would be determined by plugin location anyway), and the HelloWorld string is redundant now that the method is being passed properly. Just function pointer and stream names is enough

File: PluginSystem/PluginEngine.py [hunk 2]
```diff
@@ -8,14 +8,16 @@
 """
 Plugin Engine
 On init the engine will load all the modules inside of the plugins
-directory and add them to the list of available classes stored in PluginBase.
+directory and add them to the list of available classes stored in PluginManager.
 """
 class PluginEngine:
+    historicalCallbacks = {}
+    realtimeCallbacks = {}
```
Comment (alexweav) [inline PluginSystem/PluginEngine.py]: probably not necessary to have both. historical engine and realtime engine would ideally be entirely different classes, so no need to have both of these in one. we probably won't get to historical stuff at all, so just having one plain PluginEngine with just pluginCallbacks is probably best

File: PluginSystem/PluginEngine.py [hunk 3]
```diff
@@ -53,6 +54,29 @@ def LoadFromDirectory(self, folderPath):
...
+            # Every plugin we will pass in the engine
+            instantiate = myClass(self)
+            instantiate.init()
+
+    """
+    Registers the function and streams to the engine
+    """
+    def RegisterCallback(self, engine, name, fn, streams):
+
+        if(engine == 'realtime'):
```
Comment (alexweav) [inline PluginSystem/PluginEngine.py]: no need to make distinction see above

File: PluginSystem/PluginEngine.py [hunk 4]
```diff
@@ -53,6 +54,29 @@ def LoadFromDirectory(self, folderPath):
...
+    """
+    Returns a dictionary of all the realtimeCallbacks
+    """
+    def GetRealtimeCallbacks(self):
+        return self.realtimeCallbacks
+
+    """
+    Returns a dictionary of all the historicalCallbacks
+    """
+    def GetHistoricalCallbacks(self):
```
Comment (alexweav) [inline PluginSystem/PluginEngine.py]: same see above

File: PluginSystem/PluginEngine.py [hunk 5]
```diff
@@ -53,6 +54,29 @@ def LoadFromDirectory(self, folderPath):
...
+
+        if(engine == 'realtime'):
+            self.realtimeCallbacks[name] = [fn, streams]
+        else:
+            self.historicalCallbacks[name] = [fn, streams]
+
+    """
+    Returns a dictionary of all the realtimeCallbacks
+    """
+    def GetRealtimeCallbacks(self):
```
Comment (alexweav) [inline PluginSystem/PluginEngine.py]: GetCallbacks or GetPluginCallbacks or something

**Review by alexweav — CHANGES_REQUESTED** [review]
good stuff, but since we're low on time and probably not doing historical, theres no need to distinguish between the two for now. if we decide to do historical, it would likely be in a class anyway with its own dedicated plugin folder

File: plugins/fakePlugin.py (hunk 1 above)
Comment (mariots) [reply plugins/fakePlugin.py]: The dictionary needs a string to map to the array that holds the function pointer and the streams. So for this plugin, "HelloWorld" is the key to map to the values [self.HelloWorld, ["cameraStills", "otherStream"]]

**Comment by mariots** [comment]
@alexweav I removed the dual callback functions and replaced them with a generic GetPluginCallbacks. The only thing I didn't change was the user assigning a name to their plugin. We need the name for the dictionary.

File: plugins/fakePlugin.py (hunk 1 above)
Comment (alexweav) [reply plugins/fakePlugin.py]: function pointers are hashable in python, why not just have it be the key?

Comment (mariots) [reply plugins/fakePlugin.py]: Didn't know function pointers are hashable! I'll update it to be the key!

**Review by alexweav — APPROVED** [review]

**Pull request merged**